cassiopeia/solver/ilp_solver_utilities.pyx

```diff
@@ -1,22 +1,151 @@
 """
-Cython utilities for ILPSolver.
+Cython utilities for the ILPSolver.
 """
-cimport cython
+import cython
 
 import logging
-from typing import Dict, List, Tuple
+from typing import List, Tuple, Union
 
+# we need to doubly-import numpy so we have access to cython-optimized numpy
+# functionality
+from numpy cimport ndarray
 import numpy as np
-import pandas as pd
 
 from cassiopeia.data import utilities as data_utilities
 from cassiopeia.solver import dissimilarity_functions
 
 
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def infer_potential_graph_cython(
+    character_array: str[:, :],
```

Comment:
What's the rationale for using strings to represent characters instead of integers, like we normally do? Could we add documentation somewhere why strings are better?

Reply:
Though I originally was using integers like elsewhere, I noticed that numpy arrays of integers (and strings for that matter) take up a lot more space than simple strings. So, here I'm essentially compressing character states into character strings of the form `"char1|char2|..."` for memory efficiency. I'll make a note of this in the documentation.